File: ios/WatchAppBackend Watch App/Model/Data/Statistics/WatchData/Motion/MotionRecord.swift
//
//  MotionData.swift
//  WatchApp Watch App
//
//  Created by Mikhail Plekunov on 11/15/23.
//

import Foundation

struct MotionRecord : Codable, Equatable {
    var speed: Measurement<UnitSpeed>
    var attitude: Attitude
    var acceleration: Unit3D<UnitAcceleration>
    var gForce: Unit3D<UnitAcceleration>
    
    init(from decoder: Decoder) throws {
        let container = try decoder.container(keyedBy: CodingKeys.self)
        
        let attitude = try container.decode(Attitude.self, forKey: .attitude)
        
        let acceleration = try container.decode(Unit3D<UnitAcceleration>.self, forKey: .acceleration)
        
        let gForce = try container.decode(Unit3D<UnitAcceleration>.self, forKey: .gForce)
      
        let speed = try container.decode(Measurement<UnitSpeed>.self, forKey: .speed)
      
        self.attitude = attitude
        self.acceleration = acceleration
        self.gForce = gForce
        self.speed = speed
    }
    
    init(attitude: Attitude, speed: Measurement<UnitSpeed>, acceleration: Unit3D<UnitAcceleration>, gForce: Unit3D<UnitAcceleration>) {
        self.attitude = attitude
        self.acceleration = acceleration
        self.gForce = gForce
        self.speed = speed
    }
    
    enum CodingKeys: CodingKey {
        case attitude
        case acceleration
        case gForce
        case speed
    }
    
    func encode(to encoder: Encoder) throws {
        var container = encoder.container(keyedBy: CodingKeys.self)

        try container.encode(attitude, forKey: .attitude)
        try container.encode(acceleration, forKey: .acceleration)
        try container.encode(gForce, forKey: .gForce)
        try container.encode(speed, forKey: .speed)
    }
    
    static func == (lhs: Self, rhs: Self) -> Bool {
        return lhs.attitude == rhs.attitude &&
        lhs.acceleration == rhs.acceleration &&
        lhs.gForce == rhs.gForce &&
        lhs.speed == rhs.speed
    }
}

File: backend/model/Tracking/motionRecord.ts
import { Measurement } from "../Units/unit";
import { UnitAcceleration } from "../Units/unitAcceleration";
import { UnitAngle } from "../Units/unitAngle";
import { UnitLength } from "../Units/unitLength";
import { UnitSpeed } from "../Units/unitSpeed";

export class MotionRecord {
    speed: Measurement<UnitSpeed>
    altitude: Measurement<UnitLength>
    pitch: Measurement<UnitAngle>
    roll: Measurement<UnitAngle>
    course: Measurement<UnitAngle>
    acceleration: Measurement<UnitAcceleration>
    gForce: Measurement<UnitAcceleration>

    constructor(
        speed: Measurement<UnitSpeed>,
        altitude: Measurement<UnitLength>,
        pitch: Measurement<UnitAngle>,
        roll: Measurement<UnitAngle>,
        course: Measurement<UnitAngle>,
        acceleration: Measurement<UnitAcceleration>,
        gForce: Measurement<UnitAcceleration>
    ) {
        this.speed = speed
        this.altitude = altitude
        this.acceleration = acceleration
        this.gForce = gForce
        this.pitch = pitch
        this.roll = roll
        this.course = course
    }
}